migrator/migrations/m_192_to_m_193_report_config_migrations/migration_impl.go

```diff
@@ -0,0 +1,236 @@
+
+			continue
+		}
+
+		// create v2 report config from v1
+		newConfig := createV2reportConfig(reportConfigProto)
+
+		// create notifier config for report_configuration_notifier table
+		notifierConfig := createNotifier(reportConfigProto)
+		newConfig.Notifiers = append(newConfig.Notifiers, notifierConfig)
+
+		// add notifier to notifier_configurations_notifiers
+		reportNotifierGorm, err := updatedSchema.ConvertNotifierConfigurationFromProto(notifierConfig, 0, newConfig.GetId())
+		if err != nil {
+			errMsg := errors.Wrapf(err, "failed to convert notifier config from proto %+v", notifierConfig)
+			errList.AddError(errMsg)
+		}
+		reportNotifiers = append(reportNotifiers, reportNotifierGorm)
+		// assign version to 2 to new copy and version to 1 in original so that they are not re-created during migration
+		newConfig.Version = 2
+		reportConfigProto.Version = 1
+		// convert report config proto back to gorm model
+		convertedGormNewConfig, err := updatedSchema.ConvertReportConfigurationFromProto(newConfig)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert report config from proto %+v", newConfig)
+		}
+		convertedGormReportConfig, err := updatedSchema.ConvertReportConfigurationFromProto(reportConfigProto)
+		if err != nil {
+			errMsg := errors.Wrapf(err, "failed to convert from proto %+v", reportConfigProto)
+			errList.AddError(errMsg)
+		}
+		convertedReportConfigs = append(convertedReportConfigs, convertedGormNewConfig, convertedGormReportConfig)
+
+		// create report snapshot for last run report job
+		reportSnapshot := createReportSnapshot(reportConfigProto, newConfig)
+		if reportSnapshot != nil {
+			// convert report snapshot to GORM
+			reportSnapshotGORM, err := updatedSchema.ConvertReportSnapshotFromProto(reportSnapshot)
+			if err != nil {
+				errMsg := errors.Wrapf(err, "failed to convert report snapshot from proto %+v", reportSnapshot)
+				errList.AddError(errMsg)
+			}
+			reportSnapshots = append(reportSnapshots, reportSnapshotGORM)
+		}
+		if len(convertedReportConfigs) == batchSize {
+			err = updateTablesTransaction(tx, convertedReportConfigs, reportSnapshots, reportNotifiers)
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
+			convertedReportConfigs = convertedReportConfigs[:0]
+			reportSnapshots = reportSnapshots[:0]
+			reportNotifiers = reportNotifiers[:0]
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if rows.Err() != nil {
+		return errors.Wrap(rows.Err(), "failed to get rows for report_configurations")
+	}
+
+	err = updateTablesTransaction(tx, convertedReportConfigs, reportSnapshots, reportNotifiers)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	if !errList.Empty() {
+		log.Error(errList)
+	}
+	return tx.Commit().Error
+}
+
+func updateTablesTransaction(tx *gorm.DB, reportConfigs []*updatedSchema.ReportConfigurations, snapshots []*updatedSchema.ReportSnapshots, notifiers []*updatedSchema.ReportConfigurationsNotifiers) error {
+
+	defer func() {
```

Comment:
This doesn't go here. This goes after the call to `db.Begin()`

Reply:
@dashrews78, can we have this in the `runner.go`? Maybe a separate PR. The individual migrations should not bother about transactions.

Reply:
@md2119 IDK.  I don't think we can do the wrapping like I suggested above in `runner.go`.  Because then we would be wrapping the schema updates in a gorm transaction and that may cause scope problems for non-gorm migrations.  

That said, we could likely manually manage a transaction in `runners.go` with calls to `Begin`, `Commit` and `Rollback`.  We would probably pass the tx in the `databases` object.  Then if the migration needs a gorm transaction they use that.  But that may be a little confusing because we probably wouldn't want the schema updates in that transaction.  

We just have to think through it to make sure it is maintainable and we don't run into scope issues where something we expect was done under the transaction of some other connection mechanism.  The tough part is it is hard to refactor migrations to make things simpler because what was done still has to work.

Reply:
@md2119 I'm going to sum this up so it is less confusing.

Yes we should be able to do that.  We just have to make sure our test cases cover all the possible scenarios of gorm based vs not gorm based migrations.